color_config now accepts closures as color values

webbedspace opened the pull request:
# Description

Closes #6909. You can now add closures to your `color_config` themes. Whenever a value would be printed with `table`, the closure is run with the value piped-in. The closure must return either a {fg,bg,attr} record or a color name (`'light_red'` etc.). This returned style is used to colour the value.

This is entirely backwards-compatible with existing config.nu files.

Example code excerpt:
```
let my_theme = {
    header: green_bold
    bool: { if $in { 'light_cyan' } else { 'light_red' } }
    int: purple_bold
    filesize: { |e| if $e == 0b { 'gray' } else if $e < 1mb { 'purple_bold' } else { 'cyan_bold' } }
    duration: purple_bold
    date: { (date now) - $in | if $in > 1wk { 'cyan_bold' } else if $in > 1day { 'green_bold' } else { 'yellow_bold' } }
    range: yellow_bold
    string: { if $in =~ '^#\w{6}$' { $in } else { 'white' } }
    nothing: white
```
Example output with this in effect:
![2022-11-16 12 47 23 AM - style_computer rs_-_nushell_-_VSCodium](https://user-images.githubusercontent.com/83939/201952558-482de05d-69c7-4bf2-91fc-d0964bf71264.png)
![2022-11-16 12 39 41 AM - style_computer rs_-_nushell_-_VSCodium](https://user-images.githubusercontent.com/83939/201952580-2384bb86-b680-40fe-8192-71bae396c738.png)
![2022-11-15 09 21 54 PM - run_external rs_-_nushell_-_VSCodium](https://user-images.githubusercontent.com/83939/201952601-343fc15d-e4a8-4a92-ad89-9a7d17d42748.png)

Slightly important notes:

* Some color_config names, namely "separator", "empty" and "hints", pipe in `null` instead of a value.
* Currently, doing anything non-trivial inside a closure has an understandably big perf hit. I currently do not actually recommend something like `string: { if $in =~ '^#\w{6}$' { $in } else { 'white' } }` for serious work, mainly because of the abundance of string-type data in the world. Nevertheless, lesser-used types like "date" and "duration" work well with this.
* I had to do some reorganisation in order to make it possible to call `eval_block()` that late in table rendering. I invented a new struct called "StyleComputer" which holds the engine_state and stack of the initial `table` command (implicit or explicit).
* StyleComputer has a `compute()` method which takes a color_config name and a nu value, and always returns the correct Style, so you don't have to worry about A) the color_config value was set at all, B) whether it was set to a closure or not, or C) which default style to use in those cases.
* Currently, errors encountered during execution of the closures are thrown in the garbage. Any other ideas are welcome. (Nonetheless, errors result in a huge perf hit when they are encountered. I think what should be done is to assume something terrible happened to the user's config and invalidate the StyleComputer for that `table` run, thus causing subsequent output to just be Style::default().)
* More thorough tests are forthcoming - ran into some difficulty using `nu!` to take an alternative config, and for some reason `let-env config =` statements don't seem to work inside `nu!` pipelines(???)
* The default config.nu has not been updated to make use of this yet. Do tell if you think I should incorporate that into this.

# User-Facing Changes

See above.

# Tests + Formatting

Don't forget to add tests that cover your changes.

Make sure you've run and fixed any issues with these commands:

- `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- `cargo clippy --workspace --features=extra -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect` to check that you're using the standard code style
- `cargo test --workspace --features=extra` to check that all tests pass

# After Submitting

If your PR had any user-facing changes, update [the documentation](https://github.com/nushell/nushell.github.io) after the PR is merged, if necessary. This will help us keep the docs up to date.

fdncred commented:
fwiw, i think this is a cool change + feature! nice idea!

rgwood commented:
Haven't had a chance to look through the code, but this functionality looks very cool+useful.

zhiburt commented on crates/nu-color-config/src/nu_style.rs:
```diff
@@ -67,22 +68,29 @@ pub fn parse_nustyle(nu_style: NuStyle) -> Style {
-            },
-        };
+// Converts the color_config records, { fg, bg, attr }, into a Style.
+pub fn color_record_to_nustyle(value: Value) -> Style {
```
I think it could receive `&Value` ?

webbedspace replied:
Good point.

zhiburt commented on crates/nu-table/src/style_computer.rs:
```diff
@@ -0,0 +1,228 @@
+// It stores the engine state and stack needed to run closures that
+// may be defined as a user style.
+//
+pub struct StyleComputer<'a> {
```
Do you think it belong to `nu_table`?
I mean it could be used by many, no only for table construction, in which case there's no reason to depend on `nu_table`.

zhiburt commented on crates/nu-command/src/viewers/table.rs:
```diff
@@ -695,11 +728,12 @@ fn convert_to_table(
         .into_iter()
         .filter(|header| header != INDEX_COLUMN_NAME)
         .map(|text| {
+            let style = style_computer.compute("header", &Value::nothing(Span::unknown()));
```
Maybe to pass an actual `header`/`column` value here so the closure could be used?
*image if someone makes a closure with a string match

zhiburt replied:
Like you did bellow already actually.

webbedspace replied:
Sure.

zhiburt commented on crates/nu-command/src/viewers/table.rs:
```diff
@@ -1124,59 +1180,77 @@ fn convert_to_table2_entry(
     float_precision: usize,
 ) -> (String, TextStyle) {
     if col == 0 && with_index {
+        let style = style_computer.compute("header", &Value::nothing(Span::unknown()));
```
I think you missed here a key `row_index` must be used?

zhiburt reviewed (commented):
Went through it lightly, seems good.

Thought outloud:
Maybe we could pipe more data to the closure?
For example for tables we could provide and index (row, col), table length, etc.
So we could pass like a context.

fdncred commented on crates/nu-cli/src/repl.rs:
```diff
@@ -203,10 +201,14 @@ pub fn evaluate_repl(
-            ))
+            line_editor.with_hinter(Box::new({
+                // As of Nov 2022, "hints" color_config closures only get `null` passed in.
+                let style = style_computer.compute("hints", &Value::nothing(Span::unknown()));
```
What exactly does this mean? Are you saying `hints` won't be colored?

webbedspace replied:
What this means is, suppose you write a closure like `hints: {|e| somecommand $e }`. What should the value of $e be? It's currently just `null` because I couldn't think of a good value for it, especially since the style retrieved here gets reused for all hints. (Note that I couldn't do this for search results (in find.rs etc.) because that uses the "string" style for some reason.)

fdncred replied:
I'm sorry for being slow but I'm not understanding. I was thinking that this line meant that it's changing the default color from gray to no color. Perhaps "nothing" means just fall back to the color specified in the config?

I also thought the value of e in your closure meant that it was the current hint. so would this work `hints: {|e| $e | str upcase }` to make all hints the configured color but all uppercase?

webbedspace replied:
So, like, in the default config.nu, there's this line:
```
    hints: dark_gray
```
if you edit your config.nu and change it to
```
hints: {|e| somecommand $e }
```
then it will use whatever the output of `somecommand null` is, instead of dark_gray. If `somecommand null` returns an erroneous non-color value, then the current error handling is to just use `Style::default()`. The "

>I also thought the value of e in your closure meant that it was the current hint. so would this work hints: {|e| $e | str upcase } to make all hints the configured color but all uppercase?

The closures defined in this PR *only return colors that already were accepted by `color_config`*.
Here's all permitted types for color_config entries:

* Color name string
* { fg, bg, attr } record
* Closure -> Color name string
* Closure -> { fg, bg, attr } record

>I also thought the value of e in your closure meant that it was the current hint.

The PR currently doesn't do this because I'm concerned about the perf of running the closure every single time the hint is requested. So instead, all hints use one color. (This also applies to search results, as stated earlier).

webbedspace commented:
Did some more investigating and realised the perf isn't actually that bad after all - turns out `\w` in regexes is just really slow. 🐢　However, I still want to speed it up so that one *could* use lots of regexes if they wanted to. Had a brainwave while reading the frontmatter of https://the.exa.website/ – maybe `table` can be parallelised like `par-each`.

zhiburt commented:
> maybe table can be parallelised like par-each.

What certainly we could be parallelized is an initial width/height calculation we do for each cell.
Also `Value` to `String` conversion could be.
We certainly can't do printing.
Is there anything beside that?

webbedspace commented:
I was thinking more in terms of just the scope of this PR - parallelising the running of each closure. That could be done by shoving the StyleComputer calculation further up the line. I mean, the LS_COLORS colours are applied in a loop at the top of `handle_row_stream`… If the rest of the colour calculation was crammed up there, then it'd also fix a bug with pre-coloured LS_COLORS strings being passed into closures…

webbedspace commented:
Managed to parallelise style closure execution in convert_to_table()… had some good perf results. Still want to try and tackle the other table-making fns before I'm satisfied.

fdncred commented:
nice job parallelizing it. what type of perf results are you seeing?

webbedspace commented:
Well, using a config similar to the first post (with durations and filesizes coloured by size), I benchmarked running `ls` in a directory with 2438 subdirs, and the results were ≈958ms before → ≈570ms after.

fdncred commented:
Wow!  That's a great improvement.

zhiburt commented on crates/nu-command/src/viewers/table.rs:
```diff
@@ -878,9 +937,22 @@ fn convert_to_table(
-    let count_rows = data.len();
+    // All the computations are parallelised here.
+    // NOTE: It's currently not possible to Ctrl-C out of this...
+    let mut cells: Vec<Vec<_>> = Vec::new();
+    cells.par_extend(data.into_par_iter().map(|row| {
+        let mut new_row = Vec::new();
```
Hi there;

I wonder if you could use `Vec::with_capacity` and use index instead of `extend`.
As I see `collect_into_vec` might do exactly that?

webbedspace replied:
When I was writing those lines I was making some misassumptions about what Rayon methods would guarantee preserved ordering… I have a better sense of it now and can write this a bit different.

zhiburt commented:
(UNRELATED)

> We could parallelize an initial width/height calculation which we do for each cell.

Tested it shortly and seem like it slower on a small data set but on a bigger ones a bit faster.


|                          |  tabled          | tabled_color   | tabled_par     |
|-------------|-------------|-------------|-------------|
| test_const_table/1       |  1974.4±100.72ns | **1727.3±24.80ns** | 20.2±0.49µs    |
| test_const_table/8       |  **26.3±0.27µs**     | 27.8±0.70µs    | 79.6±4.62µs    |
| test_const_table/32      |  **395.8±2.06µs**    | 398.6±2.96µs   | 450.6±21.39µs  |
| test_const_table/128     |  **5.4±0.04ms**      | 6.1±0.04ms     | 5.5±0.28ms     |
| test_const_table/512     |  91.3±2.43ms     | 90.7±2.05ms    | **80.7±2.70ms**    |
| test_dynamic_table/1     |  1470.9±22.17ns  | **1453.8±28.48ns** | 19.3±0.53µs    |
| test_dynamic_table/8     |  22.3±0.24µs     | **21.7±0.29µs**    | 71.9±3.84µs    |
| test_dynamic_table/32    |  338.5±3.10µs    | **336.8±5.03µs**   | 401.4±21.20µs  |
| test_dynamic_table/128   |  **4.6±0.05ms**      | **4.6±0.05ms**     | 4.8±0.35ms     |
| test_dynamic_table/512   |  79.8±2.25ms     | 85.6±2.14ms    | **70.3±3.30ms**    |
| test_empty_table/1       |  **1378.8±20.19ns**  | 1408.2±11.60ns | 19.7±1.09µs    |
| test_empty_table/8       |  21.2±0.15µs     | **18.7±0.17µs**    | 66.6±4.17µs    |
| test_empty_table/32      |  **243.9±1.44µs**    | 244.0±1.50µs   | 323.0±31.64µs  |
| test_empty_table/128     |  **3.7±0.03ms**      | **3.7±0.03ms**     | 4.0±0.27ms     |
| test_empty_table/512     |  71.4±1.87ms     | 70.5±1.85ms    | **57.2±2.55ms**    |
| test_multiline_table/1   |  5.4±0.22µs      | **5.4±0.03µs**     | 29.5±0.70µs    |
| test_multiline_table/8   |  133.7±12.06µs   | **127.8±1.51µs**   | 194.4±14.10µs  |
| test_multiline_table/32  |  **1777.7±12.99µs**  | 1849.3±17.30µs | 1865.5±74.60µs |
| test_multiline_table/128 |  **28.3±0.18ms**     | **28.3±0.18ms**    | 29.1±1.86ms    |
| test_multiline_table/512 |  501.1±4.18ms    | 496.1±2.68ms   | **450.9±21.15ms**  |

PS: Just didn't know where to put it.

fdncred commented:
That pref table makes me wonder if we could do something like,
```
if rows > 256 then
  run in parallel
else
  run normally
end
```
But I think that assumes the rows are collected which isn't always the case.

zhiburt commented:
> That perf table makes me wonder if we could do something like,

Could be worth it if there will be real benefit (hard to tell, I guess need to run in nushell to see if there are any benefits).

> But I think that assumes the rows are collected which isn't always the case.

Actually we DO collect rows before the printing.
Which is actually the biggest bottle neck I am aware of.

But we basically can't eliminate it completely.
The best thing we can do is make an assumption about the width/height on a basis of some data sample (e.g. collect only 5 rows and estimate only them and truncate everything else in case it's bigger)

webbedspace commented:
Thanks for the continued support. #7059 took longer than I expected but I'll get back to this. Hmm… I guess I need to rebase this commit 718ee3d545a06 onto it… seems I'd better just do that manually.

fdncred commented:
> Starting to think I should skip further optimizations...

agreed. i'd wait a bit for more stabilization.

fdncred commented:
Having fun playing with this. :)
<img width="1723" alt="Screenshot 2022-12-11 at 6 53 39 AM" src="https://user-images.githubusercontent.com/343840/206904840-80415364-3278-4198-acef-dd2a4c56b211.png">

webbedspace commented:
OK, I've fixed up basically all the other stuff I wanted other than optimisation (namely: closure runtime errors are now printed, examples are added to default_config.nu, and StyleComputer was moved to the nu_color_config crate).

Also, I adjusted the message arrow of the "All arguments to an external command need to be string-compatible" error in run_external because I needed more information from that for testing purposes, and I figured it was a microscopic enough change to leave in.

fdncred commented on crates/nu-color-config/src/style_computer.rs:
```diff
@@ -0,0 +1,285 @@
+            initial!("record", Color::White.normal()),
+            initial!("list", Color::White.normal()),
+            initial!("block", Color::White.normal()),
+            initial!("hints", Color::DarkGray.normal()),
```
Seeing these all here made me wonder if this list was complete. So, I checked here https://github.com/nushell/nushell/blob/2bacc29d30fc6bb4798c9ef99801267492feb80d/crates/nu-parser/src/parser.rs#L2723-L2747, which may not be quite up to date either, considering this https://github.com/nushell/nushell/blob/2bacc29d30fc6bb4798c9ef99801267492feb80d/crates/nu-protocol/src/syntax_shape.rs#L9-L108.

At some point we need to get all these in sync, but we can save that for another PR.

fdncred reviewed (approved):
I'm ready to move forward with this. Thanks!

webbedspace commented:
OK, look, this is the third time I've had to manual-rebase over 7 files just to handle merge conflicts. This is getting a little ridiculous. I insist that you pull the Yes/No trigger on this in the next few days, given that since there seem to be no further comments from anyone, otherwise I'm going to start feeling like my time is being wasted.

webbedspace commented:
Everything should now be ready to merge (again) (again).

fdncred commented:
I agree 100% @webbedspace. I think we've done a miserable job with this PR, specifically around feedback to you and landing it when it was ready. I'm going to land it now. Thanks for all your support and patience.

fdncred commented:
@webbedspace I had one more thought about this that we may have to change in a new PR. There are terminals that do not support 24-bit color, namely MacOS's terminal.app. So, things like this, might need to be ported over to indexed color.
<img width="345" alt="Screenshot 2022-12-17 at 8 00 49 AM" src="https://user-images.githubusercontent.com/343840/208245711-8078f22a-ef45-42f9-8922-a1fff451a694.png">

Paying attention to only the date column, this is what it looks like in WezTerm that does support 24-bit color.
<img width="735" alt="Screenshot 2022-12-17 at 8 02 13 AM" src="https://user-images.githubusercontent.com/343840/208245733-f01816e1-508d-420c-bf92-e5c94d2eae3a.png">
The same setting in terminal.app look like this.
<img width="594" alt="Screenshot 2022-12-17 at 8 03 04 AM" src="https://user-images.githubusercontent.com/343840/208245759-801b49ac-a791-4f85-b09b-0fc11e2cbde3.png">

fdncred commented:
I tried this but it just gives an error about the theme needing to be a record.
```
  date: { (date now) - $in |
    if $in < 1hr {
      "\e[38;5;160m" #'#e61919' # 160
    } else if $in < 6hr {
      "\e[38;5;172m" '#e68019' # 172
    } else if $in < 1day {
      "\e[38;5;184m" '#e5e619' # 184
    } else if $in < 3day {
      "\e[38;5;112m" '#80e619' # 112
    } else if $in < 1wk {
      "\e[38;5;40m" '#19e619' # 40
    } else if $in < 6wk {
      "\e[38;5;44m" '#19e5e6' # 44
    } else if $in < 52wk {
      "\e[38;5;32m" '#197fe6' # 32
    } else { 'light_gray' }
  }
```
I also tried `$"(ansi idx_fg)160m"` and it didn't like that either. Maybe we don't have a good way to pass indexed colors?

webbedspace commented:
I actually think adding all the xterm colours would be a good idea in itself (mainly because they are already relatively well-known via CSS3).

fdncred commented:
I agree 100%.

I just figured out that nu-ansi-term supports these indexed colors as `Fixed`. So, we should be able to do something like this
```
    AnsiCode{ short_name: Some("red3"), long_name: "red3", code: Color::Fixed(160).prefix().to_string()},
```
although I haven't tested it yet.

fdncred commented:
Maybe you can have better luck than I. Something is not working with this and I'm not sure why. I get this when I add it to the ansi command and do `ansi --list`.
<img width="894" alt="Screenshot 2022-12-17 at 8 55 46 AM" src="https://user-images.githubusercontent.com/343840/208247929-acc11dee-d997-4bd8-a396-9cb2322e3955.png">

but when I change all the date to be all `red3` like this, it doesn't work. any ideas?
```
  date: { (date now) - $in |
    if $in < 1hr {
      'red3' #"\e[38;5;160m" #'#e61919' # 160
    } else if $in < 6hr {
      'red3' #"\e[38;5;172m" #'#e68019' # 172
    } else if $in < 1day {
      'red3' #"\e[38;5;184m" #'#e5e619' # 184
    } else if $in < 3day {
      'red3' #"\e[38;5;112m" #'#80e619' # 112
    } else if $in < 1wk {
      'red3' #"\e[38;5;40m" #'#19e619' # 40
    } else if $in < 6wk {
      'red3' #"\e[38;5;44m" #'#19e5e6' # 44
    } else if $in < 52wk {
      'red3' #"\e[38;5;32m" #'#197fe6' # 32
    } else { 'red3' }
  }
```

webbedspace commented:
Doesn't the `ansi` command use completely separate string->colour logic? Everything in nu_color_config uses `lookup_style()`.

fdncred commented:
ugh, you're right it does. oh well, i just added them all to ansi anyway. :) Let me see if I can retrofit them into lookup_style. Thanks for the tip!

fdncred commented:
aaaaarg! these xterm colors have duplicate names!

fdncred commented:
yay! got the colors mapped
<img width="562" alt="Screenshot 2022-12-17 at 10 21 04 AM" src="https://user-images.githubusercontent.com/343840/208251466-92f6e262-a778-4ad1-a38d-2528a9df6873.png">

webbedspace commented:
Just as a point of trivia, the duration-colouring example code is an approximation of a similar feature in the Windows file manager OneCommander:
![image](https://user-images.githubusercontent.com/83939/208251628-ddf10dfa-f506-4c41-b424-12995f5694dc.png)
This is the (loose) basis for the cutoff ranges I selected (1hr, 6hr, 1day etc).

fdncred commented:
it's crazy how slow this is, ` string: { if $in =~ '^#\w{6}$' { $in } else { 'white' } }`, especially in `explore`.

webbedspace commented:
It's faster if, as I said upthread, you take out `\w` and use something like `[a-fA-F\d]`.

fdncred commented:
You're right. With `explore` this is much faster. 
```
  string: { if $in =~ '^#[a-fA-F\d]+' { $in } else { 'white' } }
```

Here's some sample data just to play around.
```
['#ff0033', '#0025ee', '#0087aa', 'string', '#4101ff']
```
So, now, in `explore` I can get things like this.
<img width="665" alt="Screenshot 2022-12-21 at 10 47 05 AM" src="https://user-images.githubusercontent.com/343840/208959349-6df30a6c-294d-435b-a62e-f6d47a747825.png">

rgwood commented:
If anyone wants to optimize regex performance in these closures (and elsewhere), there is some low-hanging fruit for optimization: https://github.com/nushell/nushell/blob/046e46b962b6538b3b4b25f846a6dee48f65c1f7/crates/nu-protocol/src/value/mod.rs#L2626-L2628

I imagine we could do something like a global cache of compiled regexes, based on a FIFO queue.
